cli/sql: better error handling + display format improvements

knz opened the pull request:
Fixes  #102587.
Fixes #102606.

This PR ensures that the new ndjson format added in #77808   is included in the error message displayed when an invalid display format is configured via `\set`. It also fixes where ndjson writes its output.

It also takes the opportunity to simplify / clean up the error handling in the shell.

It also adds the output format `json` alongside `ndjson`.

Epic: None

knz committed:
clisqlexec: split the table formatting into separate files

Release note: None

knz committed:
clisqlexec,clisqlshell: auto-generate the list of table formats

This commit auto-generates the error hint about table formats from the
list of defined table formats.

Release note: None

knz committed:
cli/sql: simplify/fix error handling

Prior to this patch, error handling in the SQL shell was defective in
two ways:

- each point an error was generated was responsible for assigning
  `c.exitErr` and checking whether or not to exit the shell
  based on the `errExit` flag. This was done inconsistently.

  This patch fixes the situation by redirecting the error
  logic to a common `cliError()` function which handles
  `exitErr` and `errExit` consistently.

- the error was displayed inconsistently; some points were using
  `fmt.Fprint` and others were using `OutputError`. We want
  the latter in most cases, because it properly displays
  the hint and detail payloads.

  This patch concentrates the display of the error in the
  new `cliError()` function.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/102595)
<!-- Reviewable:end -->

knz committed:
cli/sql: fix the output for `ndjson`

Release note (bug fix): It is now possible to properly redirect the
output of SQL queries using the `ndjson` output table format in
`cockroach sql`. This bug had been introduced in v22.2.

knz committed:
cli/sql: new output format `json` alongside `ndjson`

Release note (cli change): The CLI commands that output SQL data
now support the output format JSON (`--format=json`), in addition
to newline-delimited JSON (ND-JSON, `--format=ndjson`) that had been
supported since v22.2.

knz committed:
clisqlexec: restore the test `Exampl_sql_column_labels`

The test had been mistakenly broken for a long time. This patch
restores it.

Release note: None

tbg reviewed (approved):
Nice cleanups, thank you!

otan commented on pkg/cli/clisqlexec/format_table_test.go:
```diff
@@ -136,6 +136,10 @@ thenshort`,
 	// {"a|b":"0","f\"oo":"0","f'oo":"0","f\\oo":"0","short\nvery very long\nnot much":"0","very very long\nthenshort":"0","κόσμε":"0","܈85":"0"}
+	// sql --format=json -e select * from t.u
+	// [
+	// {"a|b":"0","f\"oo":"0","f'oo":"0","f\\oo":"0","short\nvery very long\nnot much":"0","very very long\nthenshort":"0","κόσμε":"0","܈85":"0"}
```
optional: should we try prettify indent this line?

otan reviewed (approved):
did you want to backport 778094953580f76cf97aa9462e90de59f22d3c60?

rafiss reviewed (approved):
> Reviewed 1 of 1 files at r1, 8 of 9 files at r2, 1 of 3 files at r3, 8 of 8 files at r4, 2 of 2 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/102595)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz)



<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> did you want to backport [7780949](https://github.com/cockroachdb/cockroach/commit/778094953580f76cf97aa9462e90de59f22d3c60)?

Good idea, will do.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/102595#-:-NVAkjT99b4mTsNrONZP:b-e53wxc)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @rafiss)

___
*[`pkg/cli/clisqlexec/format_table_test.go` line 141 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/102595#gh-1185364583:-NVAkm-0BF9_SWf1041p:b-bk1czc) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1e448721970d70b8a1770800a14d616fb785aa3/pkg/cli/clisqlexec/format_table_test.go#L141)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

optional: should we try prettify indent this line?
</blockquote></details>

Doesn't hurt. Done.


<!-- Sent from Reviewable.io -->

knz commented:
bors r=tbg,otan,rafiss

knz commented:
bors r=tbg,otan,rafiss

knz commented:
bors r=tbg,otan,rafiss
